### ISSUE TITLE
The order of models with relationships can break DataStore on android.

### ISSUE THREAD

**Issue by TimmBaier** [aws-amplify/amplify-codegen#335]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-codegen/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-codegen/blob/master/CONTRIBUTING.md#bugs).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

v14.16.1

### Amplify CLI Version

7.6.4

### What operating system are you using?

macOS 12.0.1, android 11

### Amplify Codegen Command

codegen models

### Describe the bug

When a model has relationships (@hasMany, @belongsTo, ...), amplify codegen models, will create a different output depending on the order in which they are in the schema. This can prevent DataStore from initializing.

For example when I try to use this schema:

```graphql
type ShareProfile @model {
  id: ID!
  name: String
  profile: Profile @belongsTo
}

type Group @model {
  id: ID!
  name: String!
  profiles: [Profile] @manyToMany(relationName: "ProfileGroups")
}

type Profile @model {
  id: ID!
  name: String!
  groups: [Group] @manyToMany(relationName: "ProfileGroups")
  sharedProfiles: [ShareProfile] @hasMany
}
```

I get this error when initializing DataStore on android:

`I/amplify:aws-datastore: Creating table: PersistentModelVersion
I/amplify:aws-datastore: Creating table: Profile
I/amplify:aws-datastore: Creating table: LastSyncMetadata
I/amplify:aws-datastore: Creating table: Group
I/amplify:aws-datastore: Creating table: ShareProfile
E/SQLiteLog: (1) duplicate column name: profileSharedProfilesId in "CREATE TABLE IF NOT EXISTS `ShareProfile` (`id` TEXT PRIMARY KEY NOT NULL, `createdAt` TEXT, `name` TEXT, `profileSharedProfilesId` TEXT, `updatedAt` TEXT, `p`

If I move the model with the @belongsTo definition, below the model with the @hasMany definition for their relationship, DataStore will work without any errors.

```graphql
type Group @model {
  id: ID!
  name: String!
  profiles: [Profile] @manyToMany(relationName: "ProfileGroups")
}

type Profile @model {
  id: ID!
  name: String!
  groups: [Group] @manyToMany(relationName: "ProfileGroups")
  sharedProfiles: [ShareProfile] @hasMany
}

type ShareProfile @model {
  id: ID!
  name: String
  profile: Profile @belongsTo
}
```

`I/amplify:aws-datastore: Creating table: PersistentModelVersion
I/amplify:aws-datastore: Creating table: Profile
I/amplify:aws-datastore: Creating table: LastSyncMetadata
I/amplify:aws-datastore: Creating table: Group
I/amplify:aws-datastore: Creating table: PersistentRecord
I/amplify:aws-datastore: Creating table: ProfileGroups
I/amplify:aws-datastore: Creating table: ModelMetadata
I/amplify:aws-datastore: Creating table: ShareProfile
I/amplify:aws-datastore: Creating index for table: PersistentRecord`

### Expected behavior

`amplify codegen models` should create working code, no matter what order the models have in the schema file. Or create a warning or error, so the user knows there is something not working properly. 

### Reproduction steps

1. Create new android project and add DataStore to it
2. Create new amplify project
3. Add models with relationships (@hasOne, @hasMany, @belongsTo) to the schema, where the model with the @belongsTo definition is above the @hasOne or @hasMany definition.
4. amplify codegen models
5. The resulting java code will compile, but will not run without errors.

### GraphQL schema(s)

<details>

```graphql
# not working schema
type Group @model {
  id: ID!
  name: String!
  profiles: [Profile] @manyToMany(relationName: "ProfileGroups")
}

type Profile @model {
  id: ID!
  name: String!
  groups: [Group] @manyToMany(relationName: "ProfileGroups")
  sharedProfiles: [ShareProfile] @hasMany
}

type ShareProfile @model {
  id: ID!
  name: String
  profile: Profile @belongsTo
}
```

```graphql
# working schema
type Group @model {
  id: ID!
  name: String!
  profiles: [Profile] @manyToMany(relationName: "ProfileGroups")
}

type Profile @model {
  id: ID!
  name: String!
  groups: [Group] @manyToMany(relationName: "ProfileGroups")
  sharedProfiles: [ShareProfile] @hasMany
}

type ShareProfile @model {
  id: ID!
  name: String
  profile: Profile @belongsTo
}
```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by TimmBaier** [comment]
Here is the generated java code
[Working.zip](https://github.com/aws-amplify/amplify-codegen/files/7752880/Working.zip)
[Not_Working.zip](https://github.com/aws-amplify/amplify-codegen/files/7752889/Not_Working.zip)

This will be generated if `@belongsTo` is below `@hasMany` (this is working)
```
public final class ShareProfile implements Model {
  public static final QueryField ID = field("ShareProfile", "id");
  public static final QueryField NAME = field("ShareProfile", "name");
  public static final QueryField PROFILE = field("ShareProfile", "profileSharedProfilesId");
  private final @ModelField(targetType="ID", isRequired = true) String id;
  private final @ModelField(targetType="String") String name;
  private final @ModelField(targetType="Profile") @BelongsTo(targetName = "profileSharedProfilesId", type = Profile.class) Profile profile;
  private @ModelField(targetType="AWSDateTime", isReadOnly = true) Temporal.DateTime createdAt;
  private @ModelField(targetType="AWSDateTime", isReadOnly = true) Temporal.DateTime updatedAt;
```

and this will be generated if `@belongsTo` is above `@hasMany` (this is not working)
```
public final class ShareProfile implements Model {
  public static final QueryField ID = field("ShareProfile", "id");
  public static final QueryField NAME = field("ShareProfile", "name");
  public static final QueryField PROFILE = field("ShareProfile", "profileSharedProfilesId");
  public static final QueryField PROFILE_SHARED_PROFILES_ID = field("ShareProfile", "profileSharedProfilesId");
  private final @ModelField(targetType="ID", isRequired = true) String id;
  private final @ModelField(targetType="String") String name;
  private final @ModelField(targetType="Profile") @BelongsTo(targetName = "profileSharedProfilesId", type = Profile.class) Profile profile;
  private @ModelField(targetType="AWSDateTime", isReadOnly = true) Temporal.DateTime createdAt;
  private @ModelField(targetType="AWSDateTime", isReadOnly = true) Temporal.DateTime updatedAt;
  private final @ModelField(targetType="ID") String profileSharedProfilesId;
```

Basically it adds
`public static final QueryField PROFILE_SHARED_PROFILES_ID = field("ShareProfile", "profileSharedProfilesId");`
and
`private final @ModelField(targetType="ID") String profileSharedProfilesId;`

That is why DataStore complains that `profileSharedProfilesId` is declared twice.
```
I/amplify:aws-datastore: Creating table: ShareProfile E/SQLiteLog: (1) duplicate column name: profileSharedProfilesId in "CREATE TABLE IF NOT EXISTS ShareProfile (idTEXT PRIMARY KEY NOT NULL,createdAtTEXT,nameTEXT,profileSharedProfilesIdTEXT,updatedAtTEXT,p
```

**Comment by alharris-at** [comment]
It looks like the topological sort of our models isn't working correctly between Codegen and the Android pipeline. @lawmicha for visibility. @marcvberg to consider as part of the gql compilation rearch work.

**Comment by AaronZyLee** [comment]
@TimmBaier Thanks for bringing this out. I cannot reproduce this issue by the latest version of CLI `8.1.0`. After investigating, this was previously fixed and released after CLI version `7.6.9`. I suggest you to update the CLI version and the codegen output should stay as the same correct one no matter how the order of types in schema.
